Lokui: GUI framework experiment

Speykious opened the pull request:
This is my own GUI framework experiment.

The goal is to create an architecture out of mostly my own intuition, and see if I can reason my way through an API that makes sense.
Currently, in practice, the framework turned out to have somewhat of a retained-mode object-oriented design, but with a way clearer separation between the state and the widget tree, so probably way less opportunity to turn the GUI into a monstrous unmaintainable spaghetti.

# Explanations

## Widget tree

The widget tree is organized as a strict ownership hierarchy, so widgets cannot directly use their parents. I believe it should be possible to make any kind of interface without resorting to this kind of cyclic reference.

It persists between frames, which means that the framework is not declarative. This could potentially cause problems later down the road, as the widget tree would need to be kept manually in sync with the state in case we need to dynamically add new widgets to the tree at runtime - which is a quite common occurrence, for example a Todo List demo, or a scroll view loading more messages as we scroll up. Such a thing remains to be tested however, and it is probably manageable despite the issue.

## State

State is wrapped around either a `Lazy`, which is a newtype for `Rc<RefCell<T>>`. Since the primary goal of any kind of state is to be used and manipulated from multiple parts of the tree, it basically boils down to needing shared ownership and interior mutability. So this state management strategy is the first thing I thought of, and it turned out pretty well so far.

## Components

The most important component so far is `Pane`. It can contain other components and lay them out in different ways. Currently it can stack them on top of each other or next to each other with a flex layout, but I plan to implement a grid layout at some point too.

The other components so far, `Button` and `Text`, are just there for testing purposes. Later on, I plan to inspire myself of Flutter and Frui, to make components as small as possible to have maximum customizability. For example, a widget that reacts on a click would be or contain a `Clickable` widget, and adding some padding to a widget means wrapping it inside a `Padded` widget. This won't necessarily mean that how you build the tree itself will look extremely indented, as once we have wrappers, we just need to augment widgets with additional methods like `.on_click(...)` and `.padded(...)` in a Builder pattern way.

## Animations

> TODO

# Progress

Here are tasks I set myself to do with this framework:

- [x] Polish the layout system
- [x] Handle a basic flex layout
- [x] Render text with Skia
- [x] Add hover events
- [x] Change rectangle colors on hover for demo
- [x] Design animation system
- [ ] Create light wrapper components à la Flutter
- [ ] Complete the implementation and handling of events
- [ ] Make a Todo list UI example
- [ ] Experiment with macro for callbacks à la Appy

Speykious committed:
I think I'm on to something there

Speykious committed:
I got one button inside a panel with paddings 👀

Speykious committed:
Got events to work with the whole system too

Speykious committed:
Turn playground into counter example

Speykious committed:
Do not print while drawing

Speykious committed:
Stop printing while solving layout

Speykious committed:
Add `debug` function to `Widget` trait

Speykious committed:
Simplify `Anchor`

Speykious committed:
flex

Speykious committed:
Remove other useless prints

Speykious committed:
Get anchor-origin system right

Speykious committed:
Consistency in placement of constructor functions

Speykious committed:
Implement text rendering

Counter is now fully functional! :D

Speykious committed:
Implement `Display` for `Lazy<T>`

Speykious committed:
Implement basics of flex layout

Speykious committed:
Add a few comments

Speykious committed:
Add `debug` for `Text` component

Speykious committed:
Make `Widget::draw` take a Skia canvas

Taking lokinit's SkiaContext was stupid
since we only need the canvas anyway. :)

Speykious committed:
small format and clean

Speykious committed:
That wasn't needed

Speykious committed:
Make lokinit a dev dependency

Speykious committed:
Put buttons on true center

Useless, until the flex layout changes directions and the buttons
stay centered, probably as intended

Speykious committed:
`FlexLayout`'s `align` property has no effect

Speykious committed:
Remove `PaneChild` struct

A tuple turns out to be more convenient, I think.

Speykious committed:
Put flex layout solving into a single function

Speykious committed:
Add flex layout gap logic

Speykious committed:
refactor: functions to solve width and height

This small pattern was repeated several times.
Additionally, `default_solve_layout` is now out of the `Widget` trait
to prevent it from being overridden.

Speykious committed:
fix: cargo fmt

Speykious committed:
Handle hover events

TobiasSN commented:
Would it be okay for me to go through this and leave a few comments? I know this isn't finished so I just wanna be sure

Speykious commented:
Go ahead, would really appreciate it! I've had some itches sometimes when I look at my code so if you have any suggestions, issues or questions in general, do ask. :D

AshesOfEther reviewed (commented):
Overall, I think the ergonomics of the API could be improved substantially. It is also not clear to me how conditional UI would work, or how repeated UI like a for loop would.

I'm also concerned about the usage of `RefCell` making it easy to mutably borrow a thing in a child widget, and then trigger a panic when the parent widget tries to borrow it during the child's borrow, or vice versa.  This loses the benefit of Rust's compile-time borrow checker.

I might come back with more later, but this is what I got right now.

Speykious commented on lokui/examples/counter.rs:
```diff
@@ -0,0 +1,161 @@
+fn counter() -> Pane {
+	let value = laz(0);
+
+	let increment = {
+		let value = value.clone();
+		move |_, _| {
+			value.set(value.get() + 1);
+			println!("+1! Counter = {}", value.get());
+		}
+	};
+
+	let decrement = {
+		let value = value.clone();
+		move |_, _| {
+			value.set(value.get() - 1);
+			println!("-1! Counter = {}", value.get());
+		}
+	};
```
That's how I felt too, but I don't know of any way to make it shorter, besides a macro, which will look confusing at first glance.

TobiasSN replied:
Might be worth trying as an experiment.

Speykious replied:
Ok, I added it to the list of checks

Speykious commented on lokui/examples/counter.rs:
```diff
@@ -0,0 +1,161 @@
+	let font = lazy(Font::new(typeface, Some(20.)));
+
+	pane()
+		.with_padding(Padding::splat(10.))
+		.with_layout(
+			Layout::new()
+				.with_anchor(Anchor::CENTER)
+				.with_dimension(DimScalar::Fixed(400.), DimScalar::Fixed(250.)),
+		)
```
Probably not, as I'm gonna experiment with Flutter-style widgets in the future. Essentially these builder-style methods are gonna return their own wrapper type.

Speykious commented on lokui/examples/counter.rs:
```diff
@@ -0,0 +1,161 @@
+				.with_anchor(Anchor::CENTER)
+				.with_dimension(DimScalar::Fixed(400.), DimScalar::Fixed(250.)),
+		)
+		.child(
```
Both options should be available. Doing it the `.child(...).child(...)` way was suggested to me right after I gave an example with a `.children()` call back when the API didn't exist yet.

Speykious commented on lokui/examples/counter.rs:
```diff
@@ -0,0 +1,161 @@
+		.child(
+			pane()
+				.with_padding(Padding::vh(5., 10.))
+				.with_flex_layout(FlexLayout {
+					direction: Direction::Horizontal,
+					gap: 5.,
+				})
+				.with_layout(Layout::new().with_dimension(DimScalar::Fill, DimScalar::Fill))
```
That's true... I wonder if there's a better way... Maybe `with_flex`, not mentioning layout? Or `.flexed` to indicate that it's some kind of *flex pane*?

Speykious commented:
> I'm also concerned about the usage of `RefCell` making it easy to mutably borrow a thing in a child widget, and then trigger a panic when the parent widget tries to borrow it during the child's borrow, or vice versa.

This shouldn't happen. Components should store a `Lazy` value directly instead of a `Ref<T>` coming from a `.get()` call.
I made it so that it's intuitively what you have to do to manage shared mutable state: you `.get()`, use it and then the reference is dropped immediately; or you `.set()` and then you're already done borrowing. Since all of this happens in a single thread, there should be no mishaps as long as it's used that way.

Speykious commented:
> It is also not clear to me how conditional UI would work, or how repeated UI like a for loop would.

I don't have the answer to that yet! That's actually why I have a Todo List check, doing this example will concreticize at least one of these problems. Conditional UI should follow from there.

TobiasSN commented on lokui/src/layout/mod.rs:
```diff
@@ -0,0 +1,223 @@
+
+/// Scalar value used to represent static and dynamic widths and heights.
+#[derive(Clone, Copy, Debug, Default)]
+pub enum DimScalar {
+	/// The widget fills its parent on that dimension.
+	#[default]
+	Fill,
+	/// The widget hugs its internal content on that dimension.
+	Hug,
+	/// The dimension is fixed by that amount of pixels.
+	Fixed(f32),
+}
```
That would be the gist of it, yeah. Extending with something similar to [CSS Grid](https://developer.mozilla.org/en-US/docs/Web/CSS/CSS_Grid_Layout) later down the line, possibly making flex a subset of it, might be interesting. That's a later down the line idea though. This is a good start.

TobiasSN commented on lokui/examples/counter.rs:
```diff
@@ -0,0 +1,161 @@
+					gap: 5.,
+				})
+				.with_layout(Layout::new().with_dimension(DimScalar::Fill, DimScalar::Fill))
+				.child(
+					pane()
+						.with_layout(
+							Layout::new()
+								.with_dimension(DimScalar::Fill, DimScalar::Fixed(50.))
+								.with_origin(Anchor::CENTER)
+								.with_anchor(Anchor::CENTER),
+						)
+						.child(text(value, font.clone())),
+				)
+				.child(
+					button(text("+1", font.clone()))
+						.with_layout(
+							Layout::new()
+								.with_dimension(DimScalar::Fixed(80.), DimScalar::Fixed(50.))
+								.with_origin(Anchor::CENTER)
+								.with_anchor(Anchor::CENTER),
+						)
+						.on_click(increment),
+				)
+				.child(
+					button(text("-1", font))
+						.with_layout(
+							Layout::new()
+								.with_dimension(DimScalar::Fixed(80.), DimScalar::Fixed(50.))
+								.with_origin(Anchor::CENTER)
+								.with_anchor(Anchor::CENTER),
+						)
+						.on_click(decrement),
+				),
```
Of course, layout is what I meant to be referring to.

Speykious replied:
Ah alright

TobiasSN commented:
> This shouldn't happen. Components should store a `Lazy` value directly instead of a `Ref<T>` coming from a `.get()` call. I made it so that it's intuitively what you have to do to manage shared mutable state: you `.get()`, use it and then the reference is dropped immediately; or you `.set()` and then you're already done borrowing. Since all of this happens in a single thread, there should be no mishaps as long as it's used that way.

Ahh, I see. Thanks for the clarification!

Speykious committed:
Basics of animation system

Speykious committed:
Woops, didn't mean to change that

Speykious committed:
Add `WithBackground` widget wrapper

Speykious committed:
Add a `WidgetContainer` trait

This solves the problem of not being able to add children
after wrapping a pane with another component.

Speykious committed:
Remove `Laz`

It wasn't really useful

Speykious committed:
Remove unused imports

Speykious committed:
Use `get_mut()` in example

Speykious committed:
Remove now useless `set` method for `Lazy`

Speykious committed:
Make padding a wrapper

Speykious committed:
Slightly decrease verbosity w/ color in RectState

Speykious committed:
Rename `FlexLayout` to `Flex`

Speykious committed:
Introduce a prelude module for ease of imports

Speykious committed:
Actually making padding a wrapper was a bad idea

Speykious committed:
Add WithOnCLick and WithOnEvent wrappers

Speykious committed:
Make `on_click` callback return a boolean

Speykious committed:
Add `argb_hex` method to `Color`

Speykious committed:
Make counter example look slightly better

Speykious committed:
Replace old lokinit code with winit for now

Speykious committed:
Ah yes, the up is made out of down

Speykious committed:
Fix Skia drop order bug and simplify example
